ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java

```diff
@@ -89,138 +90,131 @@ public StructuredDataSource getSource(final String sourceName) {
-                        sourceName, sourceForQueriesMessage, sinkForQueriesMessage, sourceName));
-    }
-    dataSourceMap.remove(sourceName);
   }
 
   @Override
   public Map<String, StructuredDataSource> getAllStructuredDataSources() {
-    return dataSourceMap
+    return dataSources
         .entrySet()
         .stream()
         .collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().source));
   }
 
   @Override
   public Map<String, KsqlTopic> getAllKsqlTopics() {
-    return Collections.unmodifiableMap(topicMap);
+    return Collections.unmodifiableMap(topics);
   }
 
   @Override
-  public void updateForPersistentQuery(final String queryId,
-                                       final Set<String> sourceNames,
-                                       final Set<String> sinkNames) {
-    addSourceNames(sourceNames, queryId);
-    addSinkNames(sinkNames, queryId);
-
-  }
-
-  private void addSourceNames(final Set<String> sourceNames, final String queryId) {
-    for (final String sourceName: sourceNames) {
-      final ReferentialIntegrityTableEntry referentialIntegrity =
-          dataSourceMap.get(sourceName).referentialIntegrity;
-
-      referentialIntegrity.addSourceForQueries(queryId);
-    }
-  }
-
-  private void addSinkNames(final Set<String> sinkNames, final String queryId) {
-    for (final String sinkName: sinkNames) {
-      final ReferentialIntegrityTableEntry referentialIntegrity =
-          dataSourceMap.get(sinkName).referentialIntegrity;
-
-      referentialIntegrity.addSinkForQueries(queryId);
+  public void updateForPersistentQuery(
+      final String queryId,
+      final Set<String> sourceNames,
+      final Set<String> sinkNames
+  ) {
+    synchronized (referentialIntegrityLock) {
+      final String sourceAlreadyRegistered = streamSources(sourceNames)
+          .filter(source -> source.referentialIntegrity.getSourceForQueries().contains(queryId))
+          .map(source -> source.source.getName())
+          .collect(Collectors.joining(","));
+
+      final String sinkAlreadyRegistered = streamSources(sinkNames)
+          .filter(source -> source.referentialIntegrity.getSinkForQueries().contains(queryId))
+          .map(source -> source.source.getName())
+          .collect(Collectors.joining(","));
+
+      if (!sourceAlreadyRegistered.isEmpty() || !sinkAlreadyRegistered.isEmpty()) {
+        throw new KsqlException("query already registered."
+            + " queryId: " + queryId
+            + ", registeredAgainstSource: " + sourceAlreadyRegistered
+            + ", registeredAgainstSink: " + sinkAlreadyRegistered);
+      }
+
+      streamSources(sourceNames)
+          .forEach(source -> source.referentialIntegrity.addSourceForQueries(queryId));
+      streamSources(sinkNames)
+          .forEach(source -> source.referentialIntegrity.addSinkForQueries(queryId));
     }
   }
 
   @Override
   public void removePersistentQuery(final String queryId) {
```

Comment:
This should take the referential integrity lock too, or you might end up with a dangling query reference in one of your sources (if you have a racing updateForPersistentQuery with the same id).

Reply:
done.